Transform and state sorting reform

NolaDonato opened the pull request:
1. Added capability to calculate matrices in a general way and put them in a transform blocfk
2. Accumulated state sorting culling and transform code into a new RenderSorter class
3. Custom state sorting for main scene, post effects and shadow maps
4. Eliminated redundant GL state calls
5. Accumulated all render state into a single 64 bit number
6. Store render state in render pass
7. Added transparent flag to ShaderData
8. Added GVRRenderer class

GearVRF DCO signed off by: Nola Donato nola.donato@samsung.com

Not for merge yet - Vulkan is not working.

liaxim commented:
There are some commits adding pictures that do not belong here.

roshanch commented on GVRf/Framework/framework/src/main/jni/objects/components/render_data.cpp:
```diff
@@ -149,161 +112,20 @@ void RenderData::bindShader(JNIEnv* env, jobject localSceneObject, bool isMultiv
         render_data_string.append(std::to_string(mesh_->getVertexBuffer()->getDescriptor()));
         hash_code = render_data_string;
-        hash_code_dirty_ = false;
+        pass(0)->render_modes().clearDirty();
```
not for all the passes?

Caprica666 replied:
We have to talk about the hash code. All of the render states now fit in a single 64 bit word, one for each pass. The render passes are rendered separately, not together anymore. OpenGL doesn't use this hash code. We should discuss its significance in Vulkan.

Caprica666 commented:
I removed the one picture that was left in the repo. How do I remove commits in the middle? I was in a hurry to finish Siggraph registration and they needed photos of a specific size for the badge. I used Git to transfer stuff between PC and Linux...

liaxim commented:
You could "git revert" the unwanted commits. Or use git interactive rebase - "git rebase -i HEAD~15" to drop those commits.

liaxim commented:
And needs a rebase

liaxim commented on GVRf/Framework/framework/src/main/jni/engine/renderer/render_state.h:
```diff
@@ -0,0 +1,712 @@
+            offset_factor = src.offset_factor;
+            offset_units = src.offset_units;
+            render_order = src.render_order;
+            sample_coverage = src.sample_coverage;
```
Missing return; please compile with GVRF_USE_CLANG=true and address the warnings in the new and changed code.

liaxim commented:
gvr-accessibility doesn't look right:
![1](https://user-images.githubusercontent.com/11889541/38626692-fb63ce12-3d61-11e8-9da7-a32d2d13ba5f.png)

liaxim commented:
gvr-controls:

![2](https://user-images.githubusercontent.com/11889541/38626753-24232082-3d62-11e8-89c7-c7a371f6d5d7.png)

liaxim commented:
gvr-immersivepedia

![3](https://user-images.githubusercontent.com/11889541/38626847-5718e512-3d62-11e8-9e29-7beeb4c04ccc.png)

liaxim commented:
gvr-keyboard launches into a black screen

liaxim commented:
gvr-multilight - z-fighting?

liaxim commented:
gvr-modelviewer2 - looks weird, giant reticle

liaxim commented:
gvr-shadows doesn't look right as you look up and down

liaxim commented:
gvr-sceneobjects

![4](https://user-images.githubusercontent.com/11889541/38627133-1b5e68de-3d63-11e8-87a5-85388cd7a5a3.png)

liaxim commented:
gvr-simplephysics, gvr-switch, gvr-tutorial-lesson6, gvr-x3d-demo all don't work

liaxim commented:
Such a giant change needs to be backed up by automated tests proving correctness; or at least some level of. The previous thing was quite thoroughly tested over an extended period of time; I seriously doubt going through all the demo covers everything and can be considered enough for a change of this magnitude.

liaxim commented:
Ok, is there a Demos pull request that goes along with this one that is missing?

liaxim commented on GVRf/Framework/framework/src/main/java/org/gearvrf/shaders/GVRPhongLayeredShader.java:
```diff
@@ -101,5 +111,48 @@ protected void setMaterialDefaults(GVRShaderData material)
+     * @param isStereo       true for stereo rendering, false for mono
+     * @returns number of matrices actually calculated and stored in the output buffer
+     */
+    public static int calcMatrix(float[] inputMatrices, float[] outputMatrices, int numMatrices, boolean isStereo)
```
The calcMatrix methods on the Java side show up as unused. Are they needed?

NolaDonato commented:
PR #597 in GearVRf-Demos addresses some of these issues. I am debugging the others. I think there is an issue with render mask which is why gvr-sceneobjects is broken. I also took out the default of making everything use alpha blending. This is why some of the others are broken. We should not have this on by default.

liaxim commented on GVRf/Extensions/gvrf-particlesystem/src/main/res/raw/particle_vert.vsh:
```diff
@@ -35,14 +35,7 @@ void main() {
-       	    gl_Position =  u_mvp * posn;
-    #endif
-
+    gl_Position =  u_mvp * posn;
```
The HAS_MULTIVIEW ifdef block is really not needed anymore?

NolaDonato replied:
No, this is no longer necessary because the Java shader generation code takes care of it now.

NolaDonato commented:
There were some shader generation bugs causing some issues with the examples. I have fixed these. Now the immersivepedia, accessibility, shadows, controls, simplephysics, modelviewer2, keyboard, switch, tutorial , and x3ddemo samples work now.

roshanch commented on GVRf/Framework/framework/src/main/jni/engine/renderer/render_state.h:
```diff
@@ -0,0 +1,713 @@
+            }
+        }
+
+        uint64_t getRenderFlags() const
```
what is intention of this function?

NolaDonato replied:
This function is for sorting on render modes. It puts all the render flags in a single 64 bit integer. The only things not in there are 3 float values (polygon offset, ..)

Look at MainSceneSorter::mergeByMode to see its usage. It can also be used as a hash code.

roshanch replied:
why last 3 values are not considered for render modes? we use them in setRenderStates(). shouldn't the hash code contain all the values?

roshanch commented:
complexscene freezes.

NolaDonato commented:
This PR has been merged with master and passes all the framework tests on OpenGL. Vulkan  is still in development.

Asset tests have missing objects and sometimes only show for one eye.

The merge with master needs work.

roshanch commented:
@NolaDonato @liaxim I think, we should merge OpenGL first. Already there are many commits in this PR, so it will be better to have Vulkan support as separate PR.

NolaDonato commented:
If we do that Vulkan will be completely broken until that second pull request. In any case, take a look at the merge. Something broke in lighting and I could use some help evaluating whether I merged correctly.

liaxim commented:
It is unfortunate that this work cannot be done in more incremental way.

roshanch commented:
the reason I am saying it because efforts needed to rebase it. everyday we have to rebase it..

liaxim commented:
@roshanch I absolutely agree. This is why this work needs to be broken down and done in increments. Also this pull is of such size that it is effectively un-reviewable. Increments we can reason about and easier to test since you can tell what to target.

NolaDonato commented:
I just did the rebase. Since you have been doing it every day perhaps you can spot what I did wrong. Sushant and I are going to try to do the Vulkan part.

NolaDonato commented:
I have done a lot of testing and the problem is not a merge failure. The code works fine on a Note8 with Android O. It fails on a Note8 with Android N. It also fails on an S8 with Android O. It appears to corrupt uniform blocks. The lighting anomalies are caused by the view matrix being incorrect. Printing it right before the uniform block is loaded into the GPU shows no errors. Still investigating.

sushantojal commented:
The issue is with the GL_MAX_VERTEX_UNIFORM_COMPONENTS query. Looks like the value returned by the driver is not reliable. For gvr-multilight, any block of more than 18 matrices (288 floats) does not work for the S8+ (adreno, Android N) I am using; even though the query returns 1024 (64 mats). Works fine for values 18 and below.
[Similar](https://stackoverflow.com/questions/8586543/what-is-the-actual-number-of-vertex-uniform-components-for-glsl-shader-on-ati-gr) issues have existed for other drivers, but I wasnt able to find any documentation confirming if this is a known issue on this particular driver version.
We'll need to come up with a reliable workaround for this.

liaxim commented:
Should we close?

NolaDonato commented:
No I am going to revive this one

liaxim commented:
Revive against GearVRf instead of SXRSDK?!
